developers/dom_api.md

```diff
@@ -0,0 +1,151 @@
+duplication.
+
+<!--TODO-->
+<!--* Why not use xsd to automatically generate DOM classes-->
+
+## API Design
+
+* Each DOM class is named after a corresponding SDFormat element, e.g.,
+`sdf::World` represents `<world>` and `sdf::Model` represents `<model>`.
+
+* When parsing from an SDFormat file or string, each DOM class is responsible
+for populating its content from a given `sdf::ElementPtr` object. This is done
+in the `Load` member function of the DOM class.
+
+* The typical entry point for an end user is the `sdf::Root` object. While it's
+possible to call the `Load` member function of an object that's
+not `sdf::Root`, it is not recommended as the functionality of the object will
+be limited, (e.g., Frame semantics will not work).
+
+* If a DOM class represents an SDFormat tag that has children, then the
+class will provide member function to access the corresponding DOM objects that
+represent the children. The access can be by index or by name e.g.,
+`sdf::Model::LinkByIndex` and `sdf::Model::LinkByName`
+
+* The DOM class contains the member function `Element()` to access the
+`sdf::ElementPtr` object from which it was loaded. This is meant to provide
+access to SDFormat tags that are currently not covered by the DOM class
+including [custom tags](http://sdformat.org/tutorials?tut=custom_elements_attributes_proposal&cat=pose_semantics_docs&).
+
+* The DOM class contains mutator member functions, however, any mutation of
+a DOM object does not update the corresponding entries in the `sdf::ElementPtr`
+object contained in the DOM object
+
+* Similarly, mutation of the `sdf::ElementPtr` object of a DOM object does not
+affect the contents of the DOM object.
+
+* In general, mutation of the `sdf::ElementPtr` in a DOM object is not
+recommended as it is meant to represent the original state of the SDFormat tag
+as it was loaded from XML.
+
+### Lifetime relationships
+
+* Parent DOM objects have access to their children, but children do not have
+access to their parents. This entails that the lifetime of a parent object must
+exceed that of the child. It is invalid to use a pointer to a child object
+after the parent object has been destroyed.
+
+* Copying DOM objects makes a shallow copy of the underlying `sdf::ElementPtr`,
+thus any mutation that is done to the `sdf::ElementPtr` through one DOM objects
+affects the `sdf::ElementPtr` of the copy.
+
+* Each successfully loaded DOM object contains weak pointers to frame graphs
+that are shared by all DOM objects loaded from the same SDFormat document. The
+owner of the data is `sdf::Root`, thus, it is invalid to use frame semantics
+functions after `sdf::Root` object has been destroyed.
+
+
+### Current state of mutability
+
+* The DOM API has limited functionality for mutating DOM objects. In general,
+the API does now allow modifying existing child objects or adding or removing
+child objects.
+
+* The frame graph is only built during `sdf::Root::Load`. Subsequent updates to
```

Comment:
which version of SDFormat does this apply to? I know we changed things between 1.7 and 1.8

Reply:
Yes, this statement is true only for >= libsdformat 11.0.0. I'll update it to reflect what how it works in 9.x-10.x.